client/me/account/toggle-sites-as-landing-page.tsx

```diff
@@ -0,0 +1,53 @@
+import { ToggleControl } from '@wordpress/components';
+import { useTranslate } from 'i18n-calypso';
+import { useState } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import { successNotice } from 'calypso/state/notices/actions';
+import { savePreference } from 'calypso/state/preferences/actions';
+import { getPreference } from 'calypso/state/preferences/selectors';
+
+type ToggleSitesLandingPageProps = {
+	onToggle?: ( checked: boolean ) => void;
+	className?: string;
+};
+
+function ToggleSitesAsLandingPage( { onToggle, className }: ToggleSitesLandingPageProps ) {
+	const translate = useTranslate();
+	const dispatch = useDispatch();
+	const useSitesAsLandingPage = useSelector( ( state ) =>
+		getPreference( state, 'sites-landing-page' )
+	)?.useSitesAsLandingPage;
+
+	const [ isDisabled, setIsDisabled ] = useState( false );
+
+	async function handleToggle( isChecked: boolean ) {
+		setIsDisabled( true );
+		await dispatch(
+			savePreference( 'sites-landing-page', {
+				useSitesAsLandingPage: isChecked,
+				updatedAt: Date.now(),
+			} )
+		);
+		dispatch(
+			successNotice( translate( 'Settings saved successfully!' ), {
+				id: 'sites-landing-page-save',
+				duration: 10000,
+			} )
+		);
+		setIsDisabled( false );
```

Comment:
There's a small memory leak here. If you click the toggle and then very quickly navigate away, then by the time the network request comes back and this async function resumes the component will have unmounted. You'll see the somewhat familiar React error message `setIsDisabled` is called.

<img width="676" alt="CleanShot 2022-11-25 at 13 32 12@2x" src="https://user-images.githubusercontent.com/1500769/203878187-2516f315-6217-4c52-b0d0-0fc6ae50dfcb.png">

I'm not sure of a quick way to tidy this up. One way would be to add a `isSaving()` selector to the preferences store, that way we wouldn't need to store the `isDisabled` state in our component.
I think there might also be a way to do this by wrapping our saving logic in ReactQuery. I'll have more of a think about how that can be done, just for fun 😄 

It's unlikely the memory leak will amount to much. Users probably aren't spending all their time in Calypso toggling this setting. So I don't think it's a show-stopper.

Reply:
Here's an idea for how the leak could be fixed using ReactQuery
https://github.com/Automattic/wp-calypso/compare/add/sites-landing-page-toggle...add/sites-landing-page-toggle-react-query

The idea is that ReactQuery doesn't just have to be used to initiate network requests, it's abstractions can be used to help with anything async.

The other option I mentioned was to move the "is saving" state out of the component and into Redux. Both of these options prevent the leak in the same way—moving the state out of the React tree and thus aren't affected by components being mounted/unmounted.

Reply:
Thank you for the info on react-query, that's pretty interesting how it works in this context. Always thought it was for network requests only haha.  

I think a simpler way is to use the `savePreference` action to save to local state before issuing the call to store it remotely, with this I do not block the UI and memory leak issue is solved as I'm not using `useState` hook no more. There might be unforseen consequences if the user keeps playing with the toggle but so far I haven't encountered any.

Reply:
Yes removing the disabled state altogether also fixes it. Sometimes the disabled state is there to fix weirdness when you rapidly toggle the setting though. Like if you re-toggle it while a network request is still underway it can do weird things. But if that's not the case then it doesn't matter and we can just remove the disabled state.

Reply:
Yes like this:

https://user-images.githubusercontent.com/1500769/204163752-ea4d8920-0179-4f9d-91ff-f457c180ef91.mov

If you toggle it on and then off again quickly (e.g. if the user misclicked or they changed their mind) then when the network requests come back they cause the control to toggle. This is why we tend to disable these toggle buttons until we've finished saving.

Reply:
To be fair this isn't the only place in the UI that has this bug ...


https://user-images.githubusercontent.com/1500769/204163930-141505c1-3f88-4fa0-9cf9-6b3a8f9ba718.mov

Reply:
yeah I realise that issue and It's present alot of places in the codebase as you mentioned.  I've added the a `isSavingPreference` selector that is used to disable the button. I think it might be more helpful with the selector since it can be reused, that using the react-query way.